### PR TITLE
Rework findColumnSettingIndexForColumn to MLv2

### PR DESCRIPTION
**Pull request by uladzimirdev**
Part of  https://github.com/metabase/metabase/issues/37677

### Description

Rework to MLv2

### How to verify

Test should pass

### Demo

This is why I updated the test 
https://www.loom.com/share/7b0c28c3afdb4643b0ea56e570babc8c

### Checklist

- [x] Tests have been added/updated to cover changes in this PR

### CODE REVIEW

File: frontend/src/metabase-lib/queries/utils/dataset.ts [hunk 1]
```diff
@@ -53,21 +54,32 @@ export function findColumnIndexForColumnSetting(
...
-      _.isEqual(normalizedFieldRef, normalizeFieldRef(columnSetting.fieldRef)),
+  // ignore settings without fieldRef but preserve indexes
+  const items = columnSettings.reduce<
+    { fieldRef: FieldReference; settingIndex: number }[]
+  >((items, { fieldRef }, settingIndex) => {
+    if (fieldRef != null) {
+      items.push({ fieldRef, settingIndex });
+    }
+    return items;
+  }, []);
```
Comment (kamilmielnik) [inline frontend/src/metabase-lib/queries/utils/dataset.ts:69]: This is a bit cleaner, does not need explicit typing and helps avoid a cast later on when calling `Lib.findColumnIndexesFromLegacyRefs` (we can just use `items.map(({ fieldRef }) => fieldRef)`):

```suggestion
  const items = columnSettings.flatMap((item, settingIndex) => {
    const fieldRef = item.fieldRef ? normalizeFieldRef(item.fieldRef) : null;
    return fieldRef ? [{ fieldRef, settingIndex }] : [];
  });
```

File: frontend/src/metabase-lib/queries/utils/dataset.ts [hunk 2]
```diff
@@ -53,21 +54,32 @@ export function findColumnIndexForColumnSetting(
...
+    [column],
+    items.map(({ fieldRef }) => normalizeFieldRef(fieldRef) as FieldReference),
   );
 
-  return index;
+  const itemIndex = itemIndexes.find(index => index >= 0);
+
+  if (itemIndex != null) {
+    return items[itemIndex].settingIndex;
+  }
```
Comment (kamilmielnik) [inline frontend/src/metabase-lib/queries/utils/dataset.ts:84]: :x: Not all paths return a value

```suggestion
  }

  return -1;
```

File: frontend/src/metabase-lib/queries/utils/dataset.ts (hunk 1 above)
Comment (uladzimirdev) [reply frontend/src/metabase-lib/queries/utils/dataset.ts]: [6f121a1](https://github.com/metabase/metabase/pull/37856/commits/6f121a1bd4dab41efac91abca2c06b082578b505)

**Review by kamilmielnik — APPROVED** [review]
:+1:

**Pull request merged**